internal/rego/rego.go

```diff
@@ -264,6 +301,51 @@ func parseDirectory(directory string) ([]Rego, error) {
 	return regos, nil
 }
 
+func getBodyParamNames(rules []*ast.Rule) []string {
+	r := regexp.MustCompile(`input\.parameters\.([a-zA-Z0-9_-]+)`)
+	var bodyParams []string
+	for _, rule := range rules {
+		matches := r.FindAllStringSubmatch(rule.Body.String(), -1)
+		for _, match := range matches {
+			bodyParams = append(bodyParams, match[1]) // the 0 index is the full match, we only care about the first group
+		}
+	}
+	bodyParams = dedupe(bodyParams) // possible a param is referenced more than once
```

Comment:
suggestion: Because we don't really care about the body parameters except whether or not they exist--I would just take the header params and rewrite this to see if each header params _exists_ in the body.

maybe we could even get away with just doing a string contains of input.parameters.namehere or ..namehere[

Reply:
I disagree - the body parameters are the source of truth about what parameters the policy needs in order to work correctly. If there are any parameters in the body that aren't specified in the header comments, the generated ConstraintTemplate is incomplete and could lead to undesired policy behavior, such as allowing resources that should be blocked.

The reason we need the `@parameter` tag is to specify the type of the parameter, because we can't easily know that from the Rego itself.

Reply:
I agree with all of that, the suggestion was more on the implementation side of things. After we parse out the header comments, all we need to do is make sure they exist in the body, rather than add them into a slice then dedupe them.

The exists could probably be done with strings.Contains rather than regex since it's more existence, rather than adding all matches.

Does that make sense? Or am I missing something?

Reply:
With that route, it's possible for a user to accidentally leave out a `@parameter` tag in the comment header for a parameter that exists in the policy body. This would mean that when Gatekeeper moves to enforce the policy, the reference to the parameter would be `undefined` and could lead to unexpected consequences, such as the policy not being enforced as expected. While that is probably a rare scenario, I think we should build the tool in a way that can detect this and show an error rather than silently proceeding to generate an incomplete ConstraintTemplate. In order to do that, we must parse the parameter names from the policy body and then check if the header contains a parameter tag for each.

Reply:
Ahh ok, yes I agree. Thats a better philosophy to have. Let's roll with it.